test_mapbased_param/Cl_estimation.py

```diff
@@ -0,0 +1,375 @@
+        Cl_noise_bias.append(Cl_func(fn, fn, w)[3] )
+
+    return Cl_noise_bias
+
+
+class BBClEstimation(PipelineStage):
+    """
+    Stage that performs estimate the angular power spectra of:
+        * each of the sky components, foregrounds and CMB, as well as the cross terms
+        * of the corresponding covariance
+    """
+
+    name='BBClEstimation'
+    inputs=[('binary_mask_cut',FitsFile),('post_compsep_maps',FitsFile), ('post_compsep_cov',FitsFile),
+            ('A_maxL',NumpyFile),('noise_maps',FitsFile), ('post_compsep_noise',FitsFile), 
+            ('norm_hits_map', FitsFile), ('frequency_maps',FitsFile),('CMB_template_150GHz', FitsFile),\
+            ('mask_patches', FitsFile),('noise_cov',FitsFile)]
+    
+    outputs=[('Cl_clean', FitsFile),('Cl_noise', FitsFile),('Cl_cov_clean', FitsFile), 
+             ('Cl_cov_freq', FitsFile), ('fsky_eff',TextFile), ('Cl_fgs', NumpyFile),
+             ('Cl_CMB_template_150GHz', NumpyFile), ('mask_apo', FitsFile),
+             ('Cl_noise_bias', FitsFile)]
+
+    def run(self):
+
+        clean_map = hp.read_map(self.get_input('post_compsep_maps'),verbose=False, field=None, h=False)
+        cov_map = hp.read_map(self.get_input('post_compsep_cov'),verbose=False, field=None, h=False)
+        A_maxL = np.load(self.get_input('A_maxL'))
+        noise_maps=hp.read_map(self.get_input('noise_maps'),verbose=False, field=None)
+        post_compsep_noise=hp.read_map(self.get_input('post_compsep_noise'),verbose=False, field=None)
+        frequency_maps=hp.read_map(self.get_input('frequency_maps'),verbose=False, field=None)
+        CMB_template_150GHz = hp.read_map(self.get_input('CMB_template_150GHz'), field=None)
+        
+        nhits_raw = hp.read_map(self.get_input('norm_hits_map'))
+        nhits = hp.ud_grade(nhits_raw,nside_out=self.config['nside'])
+        nh = mknm.get_mask(nhits, nside_out=self.config['nside'])
+
+        mask_patches = hp.read_map(self.get_input('mask_patches'), verbose=False, field=None)
+        noise_cov=hp.read_map(self.get_input('noise_cov'),verbose=False, field=None)
+        # reorganization of noise covariance 
+        instrument = {'frequencies':np.array(self.config['frequencies'])}
+        ind = 0
+        noise_cov_ = np.zeros((len(instrument['frequencies']), 3, frequency_maps.shape[-1]))
+        for f in range(len(instrument['frequencies'])) : 
+            for i in range(3): 
+                noise_cov_[f,i,:] = noise_cov[ind,:]*1.0
+                ind += 1
+        # removing I from all maps
+        noise_cov_ = noise_cov_[:,1:,:]
+
+
+        nside_map = hp.get_nside(clean_map[0])
+        print('nside_map = ', nside_map)
+        
+        w=nmt.NmtWorkspace()
+        b = binning_definition(self.config['nside'], lmin=self.config['lmin'], lmax=self.config['lmax'],\
+                                         nlb=self.config['nlb'], custom_bins=self.config['custom_bins'])
+
+        print('building mask ... ')
```

Comment:
Where did we land on this one? You only need the apodization here because you're using Galactic coordinates, right? As long as you've checked the final mask has sensible derivatives it should be fine. Apodization can take annoyingly long sometimes, so you could get an apodized mask separately and just read it in here.